transformer_engine/pytorch/csrc/extensions/softmax.cu

```diff
@@ -245,3 +247,357 @@ at::Tensor scaled_aligned_causal_masked_softmax_backward(at::Tensor output_grad_
 
   return output_grads;
 }
+
+/***************************************************************************************************
+ * Support memory efficient cross entropy for Megatron-LM
+ **************************************************************************************************/
+template <typename dtype, int BlockSize>
+void __global__ CrossEntropyFwdSumExpKernel(float* sum_exp_logits_ptr,
+                                            dtype* vocab_parallel_logits_ptr, float* logits_max_ptr,
+                                            size_t n_dim) {
+  /***
+    1024 | 1
+    7 | 1016 | 2
+    6 | 1016 | 3
+
+    For example:
+    1024 | 1 -> [0,1023] [1024]
+    7 | 1016 | 2 -> [0, 6], [7,1022], [1023,1024]
+    ***/
+
+  /***
+    Thread model: size_t grid = rows;
+    One block is responsible for one row.
+    ***/
+  size_t rowIdx = blockIdx.x;
+  size_t tid = threadIdx.x;
+  if (tid >= n_dim) return;
+
+  size_t cur_vocab_parallel_logits_ptr_begin = rowIdx * n_dim;  // 0, 1025
+  size_t cur_vocab_parallel_logits_ptr_end =
+      rowIdx * n_dim + n_dim;  //cur_vocab_parallel_logits_ptr_end = 1025, 2050
+
+  size_t end_mol_num = cur_vocab_parallel_logits_ptr_end % 8;  //end_mol_num = 1, end_mol_num = 2
```

Comment:
What is `end_mol_num` (and similarly `begin_mol_num`)? Please name the variable in a way that tells the purpose of it. The comments as they are right now do not help at all and I do not understand them (why is end_mol_num 1 and then 2?).

Reply:
Yes, Sorry, the previous comments were not easy to understand. I explained the process in my response above. @ptrendx